editor/plugins/canvas_item_editor_plugin.cpp

```diff
@@ -4875,14 +4877,25 @@ void CanvasItemEditor::_popup_callback(int p_op) {
 		case SKELETON_MAKE_BONES: {
 			HashMap<Node *, Object *> &selection = editor_selection->get_selection();
 			Node *editor_root = get_tree()->get_edited_scene_root();
+			Vector<Node2D *> selected_nodes;
 
 			if (!editor_root || selection.is_empty()) {
 				return;
 			}
 
-			undo_redo->create_action(TTR("Create Custom Bone2D(s) from Node(s)"));
+			// Store the objects from `get_selection()` because otherwise the references
+			// to the selected objects would disappear, because we're repeatedly
+			// reparenting them.
 			for (const KeyValue<Node *, Object *> &E : selection) {
 				Node2D *n2d = Object::cast_to<Node2D>(E.key);
+				selected_nodes.push_back(n2d);
+			}
+
+			undo_redo->create_action(TTR("Create Custom Bone2D(s) from Node(s)"));
+			for (int i = 0; i < selected_nodes.size(); i++) {
+				undo_redo->create_action(TTR("Create Custom Bone2D from Node"));
```

Comment:
What is this action for? Hitting the `continue` below will result in invalid undo state.

Reply:
> Nested actions don't make sense if they are always nested.
That makes sense. I'll just make it one big action, then. I think I messed up when trying to rewrite the old action. Thanks!

Reply:
> What is this action for? Hitting the `continue` below will result in invalid undo state. Nested actions don't make sense if they are always nested.

The problem with how I set it up is, every action declared in an iteration in that for loop should be executed before the next iteration, so I create a new action within the loop which I execute at the end of the iteration. 

So if I were to remove the nested actions, I'd have to remove the outer action (line 4894). This would mean the `case SKELETON_MAKE_BONES` contains multiple smaller actions, and not one big 'umbrella' action. Is that acceptable? Is there a better way to do this?

Reply:
`commit_action()` does not execute the action if it's nested, so your code should have the same effect without the inner actions.

Reply:
That is weird... I've tried looking at the `undo_redo` & manager logic but this is going over my head.

In any case, as far as I understand it, what I'm trying to do shouldn't be supported. The alternative would be to pre-determine what the entire chain should look like after each node's actions would be executed, without actually executing them... I'm not sure if that's the way to go here?

Reply:
Found it. It's a bug in EditorUndoRedoManager. Your actions aren't getting nested at all.